README.md
ESTimePicker
============
[![Platform](https://cocoapod-badges.herokuapp.com/p/ESTimePicker/badge.png)](http://cocoadocs.org/docsets/ESTimePicker)
[![Version](https://cocoapod-badges.herokuapp.com/v/ESTimePicker/badge.png)](http://cocoadocs.org/docsets/ESTimePicker)

A custom time picker just like the [Google Calendar Android App](http://siliconstation.com/media/2013/06/Google-Calendar-for-Android-Time-Picker.jpg)

## Example
<br>
![Example](https://raw.github.com/e-sites/ESTimePicker/master/Assets/example.gif)


## Features

- Works both in ARC as in MRC
- Choose your own colors (background, highlight, selection, text)
- Choose your own font
- Both 24 hour notation and AM/PM
- Snap the minute view
- Completelly usable in your own viewcontroller or view

## Installation
Use cocoapods:

	pod 'ESTimePicker'
	
And then import the desired .h file:
	
	#import "ESTimePicker.h"

## Implementation
```objective-c
- (void)viewDidLoad
{
   	ESTimePicker *timePicker = [[ESTimePicker alloc] initWithDelegate:self]; // Delegate is optional
   	[timePicker setFrame:CGRectMake(10, 100, 300, 300)];
   	[self.view addSubview:timePicker];
}
	
- (void)timePickerHoursChanged:(ESTimePicker *)timePicker toHours:(int)hours
{
   	[hoursLabel setText:[NSString stringWithFormat:@"%i", hours]];
}

- (void)timePickerMinutesChanged:(ESTimePicker *)timePicker toMinutes:(int)minutes
{
   	[minutesLabel setText:[NSString stringWithFormat:@"%i", minutes]];
}
```

## Documentation
The official documentation can be found [here](http://cocoadocs.org/docsets/ESTimePicker/).

## Dependencies
This class uses the `ESMathUtils` class (which is included)


## License
Copyright (C) 2014 e-sites, [http://e-sites.nl/](http://www.e-sites.nl/). Licensed under the BSD license.


CHANGELOG.md
# ESTimePicker CHANGELOG

## 0.1.0

Initial release.


Classes/ESMathUtils.h
//
//  ESMathUtils.h
//  iOS.Library
//
//  Copyright (c) 2014, e-sites B.V. (www.e-sites.nl)
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright
//  notice, this list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright
//  notice, this list of conditions and the following disclaimer in the
//  documentation and/or other materials provided with the distribution.
//
//  Neither the name of the e-sites B.V. nor the
//  names of its contributors may be used to endorse or promote products
//  derived from this software without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
//  ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//  DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
//  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
//  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
//  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
//  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
//  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF TH
//  Created by Bas van Kuijck on 30-09-13.
//
//  [ v1.1 ]
//  - Added class method 'getDateFromSeconds'
//  - Added class method 'prefixNumberBelowTen'

#import <Foundation/Foundation.h>

@interface ESMathUtils : NSObject


+(CGFloat)degreesToRadian:(CGFloat)degrees;
+(CGFloat)radiansToDegrees:(CGFloat)radians;
+(CGFloat)distanceBetween:(CGPoint)point1 and:(CGPoint)point2;
+(CGFloat)angleBetween:(CGPoint)point1 and:(CGPoint)point2;
+(CGFloat)angleBetweenLines:(CGPoint)line1Start line1End:(CGPoint)line1End line2Start:(CGPoint)line2Start line2End:(CGPoint)line2End;
+(CGFloat)clamp:(CGFloat)val min:(CGFloat)min max:(CGFloat)max;
+(CGPoint)slopeFromDegrees:(CGFloat)degrees;
+(CGPoint)slopeFromRadians:(CGFloat)radians;
+(NSDictionary *)getDateFromSeconds:(NSTimeInterval)secs;
+(NSString *)prefixNumberBelowTen:(NSInteger)num;
@end


Classes/ESTimePicker.h
//
//  ESTimePicker.h
//  ios-library
//
//  Created by Bas van Kuijck on 20-01-14.
//
//
//  Copyright (c) 2014, e-sites B.V. (www.e-sites.nl)
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright
//  notice, this list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright
//  notice, this list of conditions and the following disclaimer in the
//  documentation and/or other materials provided with the distribution.
//
//  Neither the name of the e-sites B.V. nor the
//  names of its contributors may be used to endorse or promote products
//  derived from this software without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
//  ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//  DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
//  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
//  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
//  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
//  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
//  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

/**
 * A custom time picker just like the google calendar app on android
 */
#import <UIKit/UIKit.h>

typedef NS_ENUM(NSUInteger, ESTimePickerType) {
    ESTimePickerTypeHours,
    ESTimePickerTypeMinutes
};

@class ESTimePicker;

@protocol ESTimePickerDelegate <NSObject>
@required
- (void)timePickerHoursChanged:(ESTimePicker *)timePicker toHours:(int)hours;
- (void)timePickerMinutesChanged:(ESTimePicker *)timePicker toMinutes:(int)minutes;
@end

@interface ESTimePicker : UIView

// __________________________________________________________________________________________

/// @name Properties

// __________________________________________________________________________________________

/*!
 * The delegate of the time picker
 * @discussion You can use the time picker delegate to perform additional actions when an hour or minute is selected
 */
@property (nonatomic, assign) id<ESTimePickerDelegate> delegate;

/**
 * The color of the round wheel (aka background color)
 * @discussion Default = white
 */
@property (nonatomic, retain) UIColor *wheelColor;

/**
 * The color if a AM/PM button is pressed or the color of the middot
 * @discussion Default = blue
 */
@property (nonatomic, retain) UIColor *selectColor;

/**
 * The color of the selected hour or minute
 * @discussion Default = light blue
 */
@property (nonatomic, retain) UIColor *highlightColor;

/**
 * The color of the text used in the picker
 * @discussion Default = gray
 */
@property (nonatomic, retain) UIColor *textColor;

/**
 * The font to be used in the picker view's labels
 * @discussion Default = system font 17pt
 */
@property (nonatomic, retain) UIFont *font;

/**
 * The current selected hours
 */
@property (nonatomic, readwrite) int hours;

/**
 * The current selected minutes
 */
@property (nonatomic, readwrite) int minutes;

/**
 * Is the picker in a 24-hour format or AM/PM
 */
@property (nonatomic, readwrite, getter=isNotation24Hours) BOOL notation24Hours;

/**
 * When an hour is selected, should the pickerview automatically go to the minute view
 */
@property (nonatomic, readwrite, getter=shouldAutomaticallySwitchToMinutes) BOOL automaticallySwitchToMinutes;

/**
 * Snaps the minutes to a specific value
 * @discussion Default = 1
 */
@property (nonatomic, readwrite) unsigned int minuteSnap;

/**
 * The type of view to show
 * @discussion <pre><code>typedef NS_ENUM(NSUInteger, ESTimePickerType) {
 *      ESTimePickerTypeHours,
 *      ESTimePickerTypeMinutes
 * }</code></pre>
 */
@property (nonatomic, readwrite) ESTimePickerType type;

/**
 * Returns the current time in the correct format
 */
@property (nonatomic, readonly, assign) NSString *time;

// __________________________________________________________________________________________

/// @name Constructor

// __________________________________________________________________________________________

/**
 * Constructor
 * @param aDelegate the delegate
 * @return a ESTimePicker instance
 */
- (id)initWithDelegate:(id<ESTimePickerDelegate>)aDelegate;

// __________________________________________________________________________________________

/// @name View manipulation

// __________________________________________________________________________________________

/**
 * Changes the type
 * @param newType ESTimePickerType
 * @param animated BOOL
 */
- (void)setType:(ESTimePickerType)newType animated:(BOOL)animated;

@end

Classes/ESMathUtils.m
//
//  ESMathUtils.m
//  iOS.Library
//
//  Created by Bas van Kuijck on 30-09-13.
//
//  Copyright (c) 2014, e-sites B.V. (www.e-sites.nl)
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright
//  notice, this list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright
//  notice, this list of conditions and the following disclaimer in the
//  documentation and/or other materials provided with the distribution.
//
//  Neither the name of the e-sites B.V. nor the
//  names of its contributors may be used to endorse or promote products
//  derived from this software without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
//  ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//  DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
//  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
//  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
//  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
//  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
//  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF TH

#import "ESMathUtils.h"

@implementation ESMathUtils

+(CGFloat)degreesToRadian:(CGFloat)degrees
{
    return M_PI * degrees / 180.0;
}

+(CGFloat)radiansToDegrees:(CGFloat)radians
{
    return 180.0 * radians / M_PI;
}

+(CGFloat)distanceBetween:(CGPoint)point1 and:(CGPoint)point2
{
	return sqrt(pow(point2.x - point1.x, 2) + pow(point2.y - point1.y, 2));
}

+(CGFloat)angleBetween:(CGPoint)point1 and:(CGPoint)point2
{
	CGFloat height = point2.y - point1.y;
	CGFloat width = point1.x - point2.x;
	return [self.class radiansToDegrees:atanf(height / width)];
}

+(CGFloat)angleBetweenLines:(CGPoint)line1Start line1End:(CGPoint)line1End line2Start:(CGPoint)line2Start line2End:(CGPoint)line2End
{
	CGFloat a = line1End.x - line1Start.x;
	CGFloat b = line1End.y - line1Start.y;
	CGFloat c = line2End.x - line2Start.x;
	CGFloat d = line2End.y - line2Start.y;
	
	CGFloat rads = acos(((a*c) + (b*d)) / ((sqrt(pow(a, 2) + pow(b, 2))) * (sqrt(pow(c, 2) + pow(d, 2)))));
	
	return [self.class radiansToDegrees:rads];
}

+(CGFloat)clamp:(CGFloat)val min:(CGFloat)min max:(CGFloat)max
{
    return fminf(max, fmaxf(min, val));
}

+(CGPoint)slopeFromDegrees:(CGFloat)degrees
{
    return [self.class slopeFromRadians:[self.class degreesToRadian:degrees]];
}

+(CGPoint)slopeFromRadians:(CGFloat)radians
{
    return CGPointMake(sinf(radians), -cosf(radians));
}

+(NSDictionary *)getDateFromSeconds:(NSTimeInterval)secs
{
    
    int weeks = 0;
    int days = 0;
    int hours = 0;
    int minutes = 0;
    int ds = 60 * 60 * 24 * 7;
    while (secs >= ds) {
        weeks++;
        secs -= ds;
    }
    ds = 60 * 60 * 24;
    while (secs >= ds) {
        days++;
        secs -= ds;
    }
    ds = 60 * 60;
    while (secs >= ds) {
        hours++;
        secs -= ds;
    }
    ds = 60;
    while (secs >= ds) {
        minutes++;
        secs -= ds;
    }
    
    return @{
             @"weeks": [NSNumber numberWithInteger:weeks],
             @"days": [NSNumber numberWithInteger:days],
             @"hours": [NSNumber numberWithInteger:hours],
             @"minutes": [NSNumber numberWithInteger:minutes],
             @"seconds": [NSNumber numberWithInteger:secs]
             };
}

+(NSString *)prefixNumberBelowTen:(NSInteger)num
{
    return [NSString stringWithFormat:@"%@%i", (num<10?@"0":@""), num];
}
@end
